File: problems/0155/0155.cpp
#include <vector>

class MinStack {
private:
    std::vector <std::pair <int, int>> stack_;

public:
    MinStack() {}

    void push(int val) {
        if (stack_.empty()) {
            stack_.emplace_back(val, val);
        } else {
            int prev_min = stack_[stack_.size() - 1].second;
            stack_.emplace_back(val, std::min(val, prev_min));
        }
    }

    void pop() {
        stack_.pop_back();
    }

    int top() {
        return stack_[stack_.size() - 1].first;
    }

    int getMin() {
        return stack_[stack_.size() - 1].second;
    }
};

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack* obj = new MinStack();
 * obj->push(val);
 * obj->pop();
 * int param_3 = obj->top();
 * int param_4 = obj->getMin();
 */

File: problems/0155/0155.py
class MinStack:
    def __init__(self):
        self.stack_ = []
        self.min_stack_ = []

    def push(self, val: int) -> None:
        if not self.stack_:
            self.stack_.append(val)
            self.min_stack_.append(val)
        else:
            self.stack_.append(val)
            self.min_stack_.append(min(self.min_stack_[-1], val))

    def pop(self) -> None:
        self.stack_.pop()
        self.min_stack_.pop()

    def top(self) -> int:
        return self.stack_[-1]

    def getMin(self) -> int:
        return self.min_stack_[-1]


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(val)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()
